cp1616/include/cp1616/cp1616_io_controller.h

```diff
@@ -0,0 +1,278 @@
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+   
+* *********************************************************************************************/
+#ifndef CP1616_IO_CONTROLLER_H
+#define CP1616_IO_CONTROLLER_H
+
+#include <ros/ros.h>
+
+//IO base headers
+#include "pniousrx.h"
+#include "pnioerrx.h"
+
+#define NUM_OF_INPUT_MODULES    4       //Fixed definitions for development purposes
+#define NUM_OF_OUTPUT_MODULES   4
+
+#define MAX_NUM_OF_INIT_ATTEMPTS 1000   //Number of attempts to initialize communication
+
+/**
+ * \brief This class defines ROS-Profinet IO Controller implementation for communication processor Siemens CP1616
+ */
+
+class Cp1616IOController
+{
+public:
+
+  /**
+   * \brief Public instance accesssor
+   */
+          static Cp1616IOController* getControllerInstance();
+
+  /**
+   * \brief Destructs an IOController object
+   */
+          ~Cp1616IOController();
+
+  /**
+   * \brief Initializes and starts IO Controller
+   *
+   * \return error_code if succeded (see pnioerrx.h for detailed description)
+   */
+          int init();
+
+  /**
+   * \brief Closes and uninitializes IO Controller
+   *
+   * \return error_code if succeded (see pnioerrx.h for detailed description)
+   */
+          int uinit();
+
+   /**
+    * \brief Initialize Controller data structures
+    */
+          void configureControllerData();
+
+   /**
+   * \brief Adds input module to Controller data structre
+   *
+   * \param input_size  input data length according to STEP7 setup
+   * \param input_start_address I address according to STEP7 setup
+   *
+   * \return 0 if input module added successfully
+   */
+          int addInputModule(unsigned int input_size, unsigned int input_start_address);
+  /**
+   * \brief Adds output module to Controller data structure
+   *
+   * \param output_size output data length according to STEP7 setup
+   * \param output_start_address Q address according to STEP7 setup
+   *
+   * \return 0 if input module added successfully
+   */
+
+          int addOutputModule(unsigned int output_size, unsigned int output_start_address);
+
+  /**
+   * \brief Reads recieved data from IO Base layer
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int updateCyclicInputData();
+
+  /**
+   * \brief Writes data to IO Base library layer for transmission
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int updateCyclicOutputData();
+
+  /**
+   * \brief Changes IO Controller mode
+   *
+   * \param requested_mode OFFLINE/OPERATE/CLEAR
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int changeAndWaitForPnioMode(PNIO_MODE_TYPE requested_mode);
```

Comment:
This function name can be shorter/simpler since I would expect this system to block until the mode has changed.